### PR TITLE
Fold `static readonly VectorX` to GT_CNS_VEC

### PR DESCRIPTION
**Pull request by EgorBo**
It seems to be not uncommon: https://grep.app/search?q=static%20readonly%20Vector

```csharp
static readonly Vector128<int> vec = Vector128.Create(42);

static uint Test() => vec.ExtractMostSignificantBits();
```
New codegen:
```asm
; Assembly listing for method StaticReadonlySimd:Test():uint
       0F100509000000       movups   xmm0, xmmword ptr [reloc @RWD00]
       0F50C0               movmskps xrax, xmm0
       C3                   ret      
RWD00  	dq	0000002A0000002Ah, 0000002A0000002Ah ;; {42,42,42,42}
```
also, it enables potential constant foldings

### CODE REVIEW

**Comment by tannergooding** [comment]
`static readonly Vector<long> v4 = new Vector<long>(new long[] { 1,2,3,4 });` is one case that is good to handle

but another good case is:
```csharp
static ReadOnlySpan<int> Values => new int[] { 1, 2, 3, 4, 5, 6, 7, 8 };
// ...
Vector<int> v = new Vector<long>(Values);
```

That is rather than the `Vector<T>` being a `static readonly`, we're creating a `Vector` from an `RVA constant`

**Comment by EgorBo** [comment]
> but another good case is: 

Ah I see, but perhaps you can point me to a use-case?

**Comment by EgorBo** [comment]
@dotnet/jit-contrib PTAL (not many diffs as we don't use this pattern, but it exists in the wild - https://grep.app/search?q=static%20readonly%20Vector including OSS game engines as it seems

**Comment by EgorBo** [comment]
@jakobbotsch @tannergooding could you please take a look

**Comment by tannergooding** [comment]
> Ah I see, but perhaps you can point me to a use-case?

It's just a variation of the same general pattern. We've been recommending users utilize `Create` directly for a couple releases now.

Since `Vector<T>` can only create from a single `scalar` or from a `T[]`/`ROSpan<T>`, there will be some users that return a "constant array" from a `ROSpan<T>` property and then constructor the `Vector<T>` from that.

There is also notably the case where of this occurring during inlining

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -3942,6 +3942,52 @@ GenTree* Compiler::impImportStaticReadOnlyField(CORINFO_FIELD_HANDLE field, CORI
...
+                    {
+#ifdef TARGET_ARM64
+                        case TYP_SIMD8:
+                        case TYP_SIMD12:
+                        case TYP_SIMD16:
+                            hwAccelerated = compExactlyDependsOn(InstructionSet_AdvSimd);
+                            break;
+#endif
+#ifdef TARGET_XARCH
+                        case TYP_SIMD12:
```
Comment (tannergooding) [inline src/coreclr/jit/importer.cpp:3964]: `case TYP_SIMD8` needs to be included to cover `Vector2`

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -3942,6 +3942,52 @@ GenTree* Compiler::impImportStaticReadOnlyField(CORINFO_FIELD_HANDLE field, CORI
...
+                        case TYP_SIMD16:
+                            // Let's not complicate the test matrix with "for floating point SSE1 is enough"
+                            hwAccelerated = compExactlyDependsOn(InstructionSet_SSE2);
+                            break;
+                        case TYP_SIMD32:
+                            hwAccelerated = compExactlyDependsOn(InstructionSet_AVX);
```
Comment (tannergooding) [inline src/coreclr/jit/importer.cpp:3970]: `compOpportunisticallyDependsOn` is probably better here.

We don't "require" the given ISA, we just light up with better codegen if it isn't there and the perf difference isn't going to be drastic for rejitted code.

File: src/coreclr/jit/importer.cpp [hunk 3]
```diff
@@ -3942,6 +3942,52 @@ GenTree* Compiler::impImportStaticReadOnlyField(CORINFO_FIELD_HANDLE field, CORI
...
+                        case TYP_SIMD12:
+                        case TYP_SIMD16:
+                            hwAccelerated = compExactlyDependsOn(InstructionSet_AdvSimd);
+                            break;
+#endif
+#ifdef TARGET_XARCH
+                        case TYP_SIMD12:
+                        case TYP_SIMD16:
+                            // Let's not complicate the test matrix with "for floating point SSE1 is enough"
+                            hwAccelerated = compExactlyDependsOn(InstructionSet_SSE2);
```
Comment (tannergooding) [inline src/coreclr/jit/importer.cpp:3967]: This one and the `AdvSimd` case shouldn't be required, we can just return a constant `true`. `SSE2` and `AdvSimd` are "baseline" ISAs and `GT_CNS_VEC` doesn't depend on HWIntrinsics to perform the load/store (it directly emits the relevant instruction instead).

File: src/coreclr/jit/importer.cpp [hunk 4]
```diff
@@ -3942,6 +3942,52 @@ GenTree* Compiler::impImportStaticReadOnlyField(CORINFO_FIELD_HANDLE field, CORI
...
+                    }
+
+                    if (hwAccelerated)
+                    {
+                        GenTreeVecCon* vec   = gtNewVconNode(simdType);
+                        simd32_t       value = {};
+                        for (size_t i = 0; i < totalSize; i++)
+                        {
+                            value.i8[i] = buffer[i];
+                        }
```
Comment (tannergooding) [inline src/coreclr/jit/importer.cpp:3984]: `memcpy`?

File: src/coreclr/jit/importer.cpp [hunk 5]
```diff
@@ -3942,6 +3942,31 @@ GenTree* Compiler::impImportStaticReadOnlyField(CORINFO_FIELD_HANDLE field, CORI
...
+                int simdWidth = getSIMDTypeSizeInBytes(fieldClsHnd);
+                if (simdWidth > 0)
+                {
+                    assert((totalSize <= 32) && (totalSize <= MaxStructSize));
+                    var_types simdType = getSIMDTypeForSize(simdWidth);
+
+// SSE2 and AdvSimd are baselines so TYP_SIMD8-16 are always there
+// for TYP_SIMD32 we need to check AVX support on XARCH
+#ifdef TARGET_XARCH
+                    bool hwAccelerated = (simdType != TYP_SIMD32) || compExactlyDependsOn(InstructionSet_AVX);
```
Comment (tannergooding) [inline src/coreclr/jit/importer.cpp:3956]: nit: We tend to prefer `Opportunistic` for things like this:
```suggestion
                    bool hwAccelerated = (simdType != TYP_SIMD32) || compOpportunisticallyDependsOn(InstructionSet_AVX);
```

The reason being that this isn't required for correct codegen, it's really just an optimization to give better codegen.

**Review by tannergooding — APPROVED** [review]

**Pull request merged**